pkg/buildkit/worker/pool.go

```diff
@@ -0,0 +1,446 @@
+		WatchTimeoutSeconds: o.watchTimeoutSeconds,
+		Log:                 o.log.WithName("statefulset-scaler"),
+	})
+
+	fs := fields.SelectorFromSet(map[string]string{"status.phase": "Running"})
+	ls := labels.SelectorFromSet(conf.PodLabels)
+	podListOptions := metav1.ListOptions{
+		LabelSelector: ls.String(),
+		FieldSelector: fs.String(),
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	wp := &workerPool{
+		ctx:             ctx,
+		cancel:          cancel,
+		log:             o.log,
+		poolSyncTime:    o.syncWaitTime,
+		podMaxIdleTime:  o.maxIdleTime,
+		uuid:            string(uuid.NewUUID()),
+		requests:        NewRequestQueue(),
+		results:         make(chan PodRequestResult),
+		podClient:       clientset.CoreV1().Pods(conf.Namespace),
+		endpointsClient: clientset.CoreV1().Endpoints(conf.Namespace),
+		podListOptions:  podListOptions,
+		serviceName:     conf.ServiceName,
+		servicePort:     conf.DaemonPort,
+		namespace:       conf.Namespace,
+		scaler:          scaler,
+	}
+
+	go wp.processRequests()
+	go wp.monitorPool()
+
+	return wp
+}
+
+func (p *workerPool) Get(ctx context.Context) (string, error) {
+	ch := make(chan PodRequestResult, 1)
+	defer close(ch)
+
+	p.requests.Enqueue(ch)
+
+	oob, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go func() {
+		pod, err := p.get(oob)
+		if oob.Err() != nil {
+			p.log.V(1).Info("OOB lease get cancelled, exiting")
+			return
+		}
+
+		p.results <- PodRequestResult{pod, err}
+		p.log.V(1).Info("OOB lease get results sent")
+	}()
+
+	result := <-ch
+	if result.err != nil {
+		return "", result.err
+	}
+
+	return p.buildEndpointURL(ctx, result.pod.Name)
+}
+
+func (p *workerPool) Release(ctx context.Context, addr string) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	p.log.Info("Parsing lease addr", "addr", addr)
+	u, err := url.ParseRequestURI(addr)
+	if err != nil {
+		return fmt.Errorf("failed to parse lease addr: %w", err)
+	}
+
+	podName := strings.Split(u.Host, ".")[0]
+
+	p.log.Info("Querying for pod", "name", podName, "namespace", p.namespace)
+	pod, err := p.podClient.Get(ctx, podName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			err = fmt.Errorf("addr %q is not allocated: %w", addr, err)
+		}
+		return err
+	}
+
+	pac, err := corev1ac.ExtractPod(pod, fieldManagerName)
+	if err != nil {
+		return fmt.Errorf("cannot extract pod config: %w", err)
+	}
+
+	pac.WithAnnotations(map[string]string{
+		expiryTimeAnnotation: time.Now().Add(p.podMaxIdleTime).Format(time.RFC3339),
+	})
+	delete(pac.Annotations, leasedAnnotation)
+	delete(pac.Annotations, managerIDAnnotation)
+
+	p.log.Info("Applying pod metadata changes", "annotations", pac.Annotations)
+	if pod, err = p.podClient.Apply(ctx, pac, metav1.ApplyOptions{FieldManager: fieldManagerName}); err != nil {
+		return fmt.Errorf("cannot update pod metadata: %w", err)
+	}
+
+	if request := p.requests.Dequeue(); request != nil {
+		request <- PodRequestResult{pod, nil}
+	}
```

Comment:
That's a good question, plus I thought a lot about this. It would be nice to keep dequeuing localized but here's why I do this:

1. You release `pod-N` and send it to `results`.
2. X minutes go pass, the pool is GC'd removing that _unleased_ pod, and that result stays on the channel.
3. Next build receives the stale `pod-N` message from the channel and tries to build.

Is the pod gone? Maybe. It's a race.